src/pyodide/internal/workers.py

```diff
@@ -149,54 +157,211 @@ def json(
-        else:
-            self.js_form_data = js.FormData.new()
+    """
+    This API follows that of https://pypi.org/project/multidict/.
+    """
 
-    def __getitem__(self, key: str) -> list[FormDataValue]:
-        return list(self.js_form_data.getAll(key))
+    def __init__(
+        self, form_data: "js.FormData | None | dict[str, FormDataValue]" = None
+    ):
+        if not form_data:
+            self._js_form_data = js.FormData.new()
+            return
 
-    def __setitem__(self, key: str, value: list[FormDataValue]):
-        self.js_form_data.delete(key)
-        for item in value:
-            self.js_form_data.append(key, item)
+        if isinstance(form_data, dict):
+            self._js_form_data = js.FormData.new()
+            for k, v in form_data.items():
+                self._js_form_data.append(k, _py_value_to_js(v))
+            return
 
-    def append(self, key: str, value: FormDataValue):
-        self.js_form_data.append(key, value)
+        if (
+            hasattr(form_data, "constructor")
+            and form_data.constructor.name == "FormData"
+        ):
+            self._js_form_data = form_data
+            return
+
+        raise TypeError("Expected form_data to be a dict or an instance of FormData")
+
+    def __getitem__(self, key: str) -> FormDataValue:
+        return _js_value_to_py(self._js_form_data.get(key))
+
+    def __setitem__(self, key: str, value: FormDataValue):
+        if isinstance(value, list):
+            raise TypeError("Expected single item in arguments to FormData.__setitem__")
+        self._js_form_data.set(key, _py_value_to_js(value))
+
+    def append(self, key: str, value: FormDataValue, filename: str | None = None):
+        self._js_form_data.append(key, _py_value_to_js(value), filename)
 
     def delete(self, key: str):
-        self.js_form_data.delete(key)
+        self._js_form_data.delete(key)
 
     def __contains__(self, key: str) -> bool:
-        return self.js_form_data.has(key)
+        return self._js_form_data.has(key)
 
     def values(self) -> Generator[FormDataValue, None, None]:
-        yield from self.js_form_data.values()
+        for val in self._js_form_data.values():
+            yield _js_value_to_py(val)
 
     def keys(self) -> Generator[str, None, None]:
-        yield from self.js_form_data.keys()
+        yield from self._js_form_data.keys()
 
     def __iter__(self):
         yield from self.keys()
 
     def items(self) -> Generator[tuple[str, FormDataValue], None, None]:
-        for item in self.js_form_data.entries():
-            yield (item[0], item[1])
+        for k, v in self._js_form_data.entries():
+            yield (k, _js_value_to_py(v))
 
     def __delitem__(self, key: str):
         self.delete(key)
 
     def __len__(self):
         return len(self.keys())
+
+    def get_all(self, key: str) -> list[FormDataValue]:
+        return list(map(lambda x: _js_value_to_py(x), self._js_form_data.getAll(key)))
```

Comment:
More idiomatic to use a list comprehension here:
```suggestion
        return [_js_value_to_py(x) for x in self._js_form_data.getAll(key)]
```